Test Failure: com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest_10.testClassLoaderUnloads

RTC 311134

```
estClassLoaderUnloads:java.io.FileNotFoundException: 2026-07-18-04:42:13:633 http://localhost:8010/metric-servlet/metricServlet
    at java.base/sun.net.[www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:2048)](http://www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:2048))
    at java.base/sun.net.[www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1641)](http://www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1641))
    at com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest.getServlet(MetricsClassLoaderTest.java:101)
    at com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest.testClassLoaderUnloads(MetricsClassLoaderTest.java:44)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:242)
    at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:412)
    at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:204)

Further analysis is available at
https://libh-proxy1.fyre.ibm.com/cognitive/defectAnalysis.html?failureGroupId=failure_group-e1ce52f1-2ece-473f-b5b7-62f5ef13d0e0

Potential root issue:

MetricsClassLoaderServer_1.0-18-07-2026-04-42-19
messages.log:

[7/18/26, 4:42:04:859 UTC] 0000002a com.ibm.ws.kernel.feature.internal.FeatureManager            A CWWKF0011I: The MetricsClassLoaderServer_1.0 server is ready to run a smarter planet. The MetricsClassLoaderServer_1.0 server started in 41.757 seconds.
[7/18/26, 4:42:13:548 UTC] 0000002e com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "com.ibm.wsspi.injectionengine.InjectionException: java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "getProtectionDomain") com.ibm.ws.webcontainer.servlet.ServletWrapper.loadServlet 228" at ffdc_26.07.18_04.42.13.0.log
[7/18/26, 4:42:13:553 UTC] 0000002e com.ibm.ws.webcontainer.servlet                              E SRVE0319E: For the [com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricServlet] servlet, com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricServlet servlet class was found, but a resource injection failure has occurred. java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "getProtectionDomain")
[7/18/26, 4:42:13:558 UTC] 0000002e com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "javax.servlet.UnavailableException: SRVE0319E: For the [com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricServlet] servlet, com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricServlet servlet class was found, but a resource injection failure has occurred. java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "getProtectionDomain") com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest() 302" at ffdc_26.07.18_04.42.13.1.log


FFDC:

------Start of DE processing------ = [7/18/26, 4:42:13:291 UTC]
Exception = com.ibm.wsspi.injectionengine.InjectionException
Source = com.ibm.ws.webcontainer.servlet.ServletWrapper.loadServlet
probeid = 228
Stack Dump = com.ibm.wsspi.injectionengine.InjectionException: java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "getProtectionDomain")
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.inject(WebApp.java:1283)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.injectAndPostConstruct(WebApp.java:1419)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.injectAndPostConstruct(WebApp.java:1407)
    at com.ibm.ws.webcontainer.osgi.servlet.ServletWrapper.createTarget(ServletWrapper.java:65)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper$1.run(ServletWrapper.java:1547)
    at java.base/java.security.AccessController.doPrivileged(AccessController.java:757)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.loadServlet(ServletWrapper.java:1515)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:595)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:448)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1362)
    at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5106)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:328)
    at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1047)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:293)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1634)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:781)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:741)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:574)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:505)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:365)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:332)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:169)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:77)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:513)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:583)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:967)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1056)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:344)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:854)
Caused by: java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "getProtectionDomain")
    at java.base/java.security.AccessController.throwACE(AccessController.java:182)
    at java.base/java.security.AccessController.checkPermissionHelper(AccessController.java:246)
    at java.base/java.security.AccessController.checkPermission(AccessController.java:393)
    at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:416)
    at java.base/java.lang.Class.getProtectionDomain(Class.java:2068)
    at com.ibm.ws.classloading.internal.AppClassLoader.getCodeSourceString(AppClassLoader.java:661)
    at com.ibm.ws.classloading.internal.AppClassLoader.definePackageAndClass(AppClassLoader.java:541)
    at com.ibm.ws.classloading.internal.AppClassLoader.findClass(AppClassLoader.java:415)
    at com.ibm.ws.classloading.internal.AppClassLoader.findOrDelegateLoadClass(AppClassLoader.java:751)
    at com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:632)
    at com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:601)
    at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1105)
    at com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricGenerator.<init>(MetricGenerator.java:81)
    at com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricGenerator$Proxy$_$$_WeldClientProxy.<init>(Unknown Source)
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
    at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
    at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
    at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:347)
    at java.base/java.lang.Class.newInstance(Class.java:2420)
    at org.jboss.weld.bean.proxy.DefaultProxyInstantiator.newInstance(DefaultProxyInstantiator.java:43)
    at org.jboss.weld.bean.proxy.ProxyFactory.run(ProxyFactory.java:362)
    at java.base/java.security.AccessController.doPrivileged(AccessController.java:701)
    at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:346)
    at org.jboss.weld.bean.proxy.ClientProxyFactory.create(ClientProxyFactory.java:81)
    at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:206)
    at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:196)
    at org.jboss.weld.bean.proxy.ClientProxyProvider.access$100(ClientProxyProvider.java:44)
    at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:52)
    at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:48)
    at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache$1.apply(ReentrantMapBackedComputingCache.java:55)
    at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache$1.apply(ReentrantMapBackedComputingCache.java:51)
    at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getValue(ReentrantMapBackedComputingCache.java:64)
    at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getCastValue(ReentrantMapBackedComputingCache.java:80)
    at org.jboss.weld.bean.proxy.ClientProxyProvider.getClientProxy(ClientProxyProvider.java:230)
    at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:779)
    at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:885)
    at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
    at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:358)
    at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:369)
    at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:70)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl$3.run(WebSphereInjectionServicesImpl.java:325)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl$3.run(WebSphereInjectionServicesImpl.java:322)
    at java.base/java.security.AccessController.doPrivileged(AccessController.java:701)
    at com.ibm.ws.cdi.impl.weld.injection.WebSphereInjectionServicesImpl.aroundInject(WebSphereInjectionServicesImpl.java:322)
    at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:46)
    at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:72)
    at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.cdiInjection(CDIManagedObject.java:204)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.inject(CDIManagedObject.java:177)
    at com.ibm.ws.cdi.impl.managedobject.CDIManagedObject.inject(CDIManagedObject.java:155)
    at com.ibm.ws.webcontainer.osgi.webapp.WebApp.inject(WebApp.java:1278)
```

--- a/dev/com.ibm.ws.microprofile.metrics_fat/publish/servers/MetricsClassLoaderServer_1.0/server.xml
+++ b/dev/com.ibm.ws.microprofile.metrics_fat/publish/servers/MetricsClassLoaderServer_1.0/server.xml
@@ -28,5 +28,10 @@
     <webApplication id="metric-servlet" location="metric-servlet.war" name="metric-servlet">
     	<classloader commonLibraryRef="lib"/>
     </webApplication> 
+    
+    <javaPermission
+    codeBase="${server.config.dir}/apps/metric-servlet.war"
+    className="java.lang.RuntimePermission"
+    name="getProtectionDomain"/>
 
 </server>

